pkg/diagnostics/workflow_monitoring.go

```diff
@@ -0,0 +1,121 @@
+	"go.opencensus.io/tag"
+
+	diagUtils "github.com/dapr/dapr/pkg/diagnostics/utils"
+)
+
+var executionTypeKey = tag.MustNewKey("execution_type")
+
+const (
+	StatusSuccess     = "success"
+	StatusFailed      = "failed"
+	StatusRecoverable = "recoverable"
+	CreateWorkflow    = "create_workflow"
+	GetWorkflow       = "get_workflow"
+	AddEvent          = "add_event"
+	PurgeWorkflow     = "purge_workflow"
+	Activity          = "activity"
+	Workflow          = "workflow"
+
+	WorkflowEvent = "event"
+	Timer         = "timer"
+
+	ComponentName = "dapr"
+)
+
+type workflowMetrics struct {
+	// workflowOperationCount records count of Successful/Failed requests to Create/Get/Purge Workflow and Add Events.
+	workflowOperationCount *stats.Int64Measure
+	// workflowOperationLatency records latency of response for workflow operation requests.
+	workflowOperationLatency *stats.Float64Measure
+	// workflowExecutionCount records count of Successful/Failed/Recoverable workflow/activity executions.
+	workflowExecutionCount *stats.Int64Measure
+	// workflowExecutionLatency records latency of workflow/activity executions.
+	workflowExecutionLatency *stats.Float64Measure
+	appID                    string
+	enabled                  bool
+	namespace                string
+}
+
+func newWorkflowMetrics() *workflowMetrics {
+	return &workflowMetrics{
+		workflowOperationCount: stats.Int64(
+			"runtime/workflow/operation/count",
+			"The number of successful/failed workflow operation requests.",
+			stats.UnitDimensionless),
+		workflowOperationLatency: stats.Float64(
+			"runtime/workflow/operation/latency",
+			"The latencies of responses for workflow operation requests.",
+			stats.UnitMilliseconds),
+		workflowExecutionCount: stats.Int64(
+			"runtime/workflow/execution/count",
+			"The number of successful/failed/recoverable workflow/activity executions.",
+			stats.UnitDimensionless),
+		workflowExecutionLatency: stats.Float64(
+			"runtime/workflow/execution/latency",
+			"The total time taken to run a workflow/activity to completion.",
+			stats.UnitMilliseconds),
+	}
+}
+
+func (w *workflowMetrics) IsEnabled() bool {
+	return w != nil && w.enabled
+}
+
+// Init registers the workflow metrics views.
+func (w *workflowMetrics) Init(appID, namespace string) error {
+	w.appID = appID
+	w.enabled = true
+	w.namespace = namespace
+
+	return view.Register(
+		diagUtils.NewMeasureView(w.workflowOperationCount, []tag.Key{appIDKey, componentKey, namespaceKey, operationKey, statusKey}, view.Count()),
+		diagUtils.NewMeasureView(w.workflowOperationLatency, []tag.Key{appIDKey, componentKey, namespaceKey, operationKey, statusKey}, defaultLatencyDistribution),
+		diagUtils.NewMeasureView(w.workflowExecutionCount, []tag.Key{appIDKey, componentKey, namespaceKey, executionTypeKey, statusKey}, view.Count()),
+		diagUtils.NewMeasureView(w.workflowExecutionLatency, []tag.Key{appIDKey, componentKey, namespaceKey, executionTypeKey, statusKey}, defaultLatencyDistribution))
```

Comment:
I think we should have workflowName and activityName as well in this metric; else a common plotting won't make much sense - as different activities mayhave different expected latencies.

Reply:
Sure

Reply:
Done